[ENH] consolidating splitters as their own module with systematic tests and extension

BensHamza opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/sktime/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #4889 

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
- [x] Create a new module, `sktime.split`, which will contain all the splitters
- [x] Create submodule `base` which only has the `BaseSplitter` and `BaseWindowSplitter`
- [x] Move the splitters into individual files inside `split`. `__init__` should export them all. The old location should be subject to deprecation - imports should still work but raise a warning.
- [x] Create a test module, `split.tests`. All current tests should move there.
- [ ] Add an extension template for splitters, in `extension_templates` - this will inform tests below as it will make concrete the key methods.
- [ ] Create a new test class, `tests::TestAllSplitters`. This should inherit from `TestAllObjects`, and be similar to `TestAllAligners`, `TestAllDistributions`, and similar. It should also be registered in `check_estimators` (see there for pattern)
- [ ] The generic tests should be migrated to `TestAllObjects`, as long as it is a test for all splitters.
- [ ] Think about new tests, e.g., "test that `get_n_splits` is always the same as length of `split`". All methods of splitters should be covered.
- [ ] It may make sense to add test scenarios, although this may equally be covered by simple fixtures.

#### Does your contribution introduce a new dependency? If yes, which one?
No
<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?
Still no test
<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->
This work is still on progress, but open to discussions, remarks and reviews.

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).


<!--
Thanks for contributing!
-->

BensHamza commented:
Thank you @hazrulakmal for your comment. Indeed I pushed the code without `pre-commit` configured. I make the changes right now.

fkiraly commented on sktime/forecasting/model_selection/_split.py:
```diff
@@ -67,6 +68,12 @@
 SPLIT_GENERATOR_TYPE = Iterator[SPLIT_ARRAY_TYPE]
 PANDAS_MTYPES = ["pd.DataFrame", "pd.Series", "pd-multiindex", "pd_multiindex_hier"]
 
+warnings.warn(
```
if you want to avoid that this warning is raised when running crawling utility `all_estimator`, this should be added to the excepted module list there.

Also, replace "this module" by the full name, as users receiving the warning from an indirect import may get confused.

BensHamza replied:
Thank you for the feedback!
I've added the `_split` and `test_split` modules to the `EXCEPTED_MODULES` list in the `all_estimator` crawling utility as suggested and I've updated the deprecation warning to use the full module name using the `"__name__"` attribute for clarity.
Could you please review the updated changes?

hazrulakmal commented:
Notes: some other PRs that may affect this PR.

1. #5012 
2. #5009

fkiraly commented:
Had a quick look - the failing test `test_inheritance` seems failing as the check fails that the splitters inherit from `BaseSplitter`.

I suppose this is due to a duplication of classes - the registry should contain the correct class, and the old class should be removed.

BensHamza commented:
> Had a quick look - the failing test `test_inheritance` seems failing as the check fails that the splitters inherit from `BaseSplitter`.
> 
> I suppose this is due to a duplication of classes - the registry should contain the correct class, and the old class should be removed.

Thank you @fkiraly for identifying the issue. I made the changes in `registry`.

fkiraly commented:
hm, it's still not identifying the inheritance tree correctly - one of the two sides on the comparison must be mismatched.

I would advise, try to run the test manually and see how the mismatch comes to pass. I can't grok from naive inspection what's the root cause. If you can't find it, I'm happy to check in a screenshare in the Fri meet-up or mentoring meeting, if you can make it this week.

BensHamza commented:
I will try tomorrow before the meeting but I think your help is really appreciated here @fkiraly, thank you ! 😁

fkiraly commented:
ok, let's look at it later in a screenshare.
I would also advise, let's cut the plan up in multiple PR, and try to merge the "move-refactor" soon, and look at the test class based on that

fkiraly commented:
from discussion on Sep 8:

* splitters are currently duplicated, in the old and the new module
* that is the deeper reason behind the test failures
* instead, the old module should be removed, with imports pointing to the new module so users can still import from the old location
* a warning should be raised when a user imports from the old location, pointing them to the new import location
* deprecation horizon should be one minor cycle ahead, right now that would be 0.24.0
* a todo comment should be left in the old module that starts with `# todo 0.24.0`, with explicit instructions what the release manager should do at 0.24.0 (if we merge after 0.23.0, this needs to change accordingly)

fkiraly reviewed (changes requested):
Great, I looked through this - almost done!

One thing I´d still like to ask, we need to make sure that we do not break code of downstream users, who may still import splitters from `sktime.forecasting.model_selection`.

For that, we need to follow the typical deprecation cycle:

* until 0.25.0, allow imports from that location, but still raise a warning. This can be achieved by the old location importing from the new location.
* at 0.25.0, remove the old location.
* leave `todo 0.25.0` messages in the code that tells the release manager for 0.25.0 which files or content to remove.

Let me know if you need further details!

BensHamza commented:
> Great, I looked through this - monumental work! We're almost done!

Thank you for your enormous help and dedication ! I learned a lot trough the process.

> * until 0.25.0, allow imports from that location, but still raise a warning.

Importing splitters from  `sktime.forecasting.model_selection` or `sktime.forecasting.model_selection._split` works. the deprecation warning is also raised for both.

> * leave `todo 0.25.0` messages in the code that tells the release manager for 0.25.0 which files or content to remove.

I left the todo comment for the `sktime.forecasting.model_selection._split` module, the only module that is affected by . The old location of splitters' test is deleted (transferred with adaptation to `sktime.split.tests`) in this PR.

fkiraly commented:
> the deprecation warning is also raised for both.

Great! The warning should say though that the import will stop working in 0.25.0.

> I left the todo comment 

I see! For search purposes, the comment should start literally with `todo 0.25.0`.

BensHamza commented:
> The warning should say though that the import will stop working in 0.25.0.

I added it in my last commit, here is the actual message : 
```
Please import sktime splitters from `sktime.split`. Importing splitters from `sktime.forecasting.model_selection` is deprecated and will be removed in release 0.25.0
```
> the comment should start literally with todo 0.25.0.

Sorry I didn't mention it correctly. I added the comment `todo 0.25.0 : Please ensure this file is deleted.` in the specified deprecated file with a quick description above it.

I can change the message or else if needed. Thank you !

fkiraly reviewed (approved):
Excellent, thanks! Now it is very clear!
